presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,232 @@
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        {
+            if (!SystemSessionProperties.isOptimizeNullsInJoinv2(session)) {
+                return joinNode;
+            }
+
+            List<JoinNode.EquiJoinClause> joinCriteria = joinNode.getCriteria();
+            RowExpression joinFilter = joinNode.getFilter().orElse(null);
+
+            final ArrayList<VariableReferenceExpression> leftExpressions = new ArrayList<>();
+            final ArrayList<VariableReferenceExpression> rightExpressions = new ArrayList<>();
+
+            //For equi-join clause, we have already resolved variable references
+            for (JoinNode.EquiJoinClause clause : joinCriteria) {
+                leftExpressions.add(clause.getLeft());
+                rightExpressions.add(clause.getRight());
+            }
+
+            if (joinFilter != null) {
+                //Parse the joinFilter and extract variable references that can be inferred to be NOT NULL
+                List<VariableReferenceExpression> refs = extractInferredNotNullVariables(joinFilter);
+
+                final List<VariableReferenceExpression> leftOutputVars = joinNode.getLeft().getOutputVariables();
+                final List<VariableReferenceExpression> rightOutputVars = joinNode.getRight().getOutputVariables();
+
+                for (VariableReferenceExpression ref : refs) {
+                    if (leftOutputVars.contains(ref)) {
+                        leftExpressions.add(ref);
+                    }
+                    else if (rightOutputVars.contains(ref)) {
+                        rightExpressions.add(ref);
+                    }
+                    else {
+                        throw new RuntimeException(String.format("Variable references extracted did not exist in left or right output variables : [%s]", ref));
+                    }
+                }
+            }
+
+            switch (joinNode.getType()) {
+                case INNER:
+                    //Merge the lists, we need to add NOT NULL filters for both sides
+                    leftExpressions.addAll(rightExpressions);
+                    return newJoinNode(context, joinNode, combinedFilter(joinFilter, buildNotNullRowExpression(leftExpressions)));
+                case LEFT:
+                    //Right side not null filters can be added
+                    return newJoinNode(context, joinNode, combinedFilter(joinFilter, buildNotNullRowExpression(rightExpressions)));
+                case RIGHT:
+                    //Left side not null filters can be added
+                    return newJoinNode(context, joinNode, combinedFilter(joinFilter, buildNotNullRowExpression(leftExpressions)));
+                case FULL:
+                default:
+                    return newJoinNode(context, joinNode, joinFilter);
+            }
+        }
+
+        private List<VariableReferenceExpression> extractInferredNotNullVariables(RowExpression joinFilter)
```

Comment:
I am a bit concerned about this optimization, looks aggressive to me, for example it depends on the return value of isCalledOnNullInput which I am not 100% confident about.

Reply:
Can you elaborate on your concerns ?  
The property here is populated from a function's [isCalledOnNullInput](https://github.com/prestodb/presto/blob/d09e4df79f5b7c122c70dca8d080f5be423ee59d/presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunction.java#L24) so I was assuming this to be 'safe'

Reply:
I mean I am not 100% confident if `isCalledOnNullInput` is correctly set for all functions.
The optimization based on InferredNotNullVariables looks highly risky to me. At least I think it's worth a separate session property to control it, and add some comments on it.

Also please implement this optimization as an iterative optimization rule.

Reply:
> I mean I am not 100% confident if `isCalledOnNullInput` is correctly set for all functions. The optimization based on InferredNotNullVariables looks highly risky to me. At least I think it's worth a separate session property to control it, and add some comments on it.
> 
> Also please implement this optimization as an iterative optimization rule.

Yeah this is pretty tricky to get right and also do you check for expression determinism? I say it's best to stick to equality conditions for simplicity.

Reply:
>Also please implement this optimization as an iterative optimization rule
I have concerns with the feasibility of making this an iterative optimization rule. Taking this offline on Slack

Reply:
@feilong-liu 
### On using `isCalledOnNullInput`

The change to add isCalledOnNullInput was done in [PR-12570](https://github.com/prestodb/presto/pull/12570) , in commit[4a071818](https://github.com/prestodb/presto/pull/12570/commits/4a071818b034793d9eeae9ca158538026dc9a049). The core change here is to replace calls like `function.getArgumentProperty(i).getNullConvention() == RETURN_NULL_ON_NULL` with `functionMetadata.isCalledOnNullInput()` (see e.g `RowExpressionInterpreter`)
[This comment](https://github.com/prestodb/presto/pull/12570#issuecomment-481847477) calls out relying on test failures to check if this property is not set correctly

Additional data points :

1. For [BuiltInFunction](https://github.com/prestodb/presto/blob/44ef3931becfc2b334e4710224e2dcb18a19c12d/presto-main/src/main/java/com/facebook/presto/metadata/BuiltInFunction.java#L18)'s the default is `false`.  

2. I ran a tweaked `SHOW FUNCTIONS`  that adds the output of `isCalledOnNullInput()`. I did this for the full list of built in functions + operators. Results are on [this doc](https://docs.google.com/spreadsheets/d/1H3Q11HqK8k9AWOUP_jELISttdPUQsL58SWCvWPyKQxY/edit?usp=sharing). 
From my analysis, majority of the functions have this set to  `false` (the default), the ones that have it `true` I can see that it make sense to have them there

IMO this gives us fair amount of confidence that this property can be relied on..

Reply:
@feilong-liu @ClarenceThreepwood @kaikalur 
## On rewriting this as an Iterative optimizer

I had experimented with writing this optimization as a iterative optimizer, but decided against it.

1.  This optimization can be run on any JoinNode, i.e the pattern to match is just a plain `join()`
2.  The results of the optimzation in this PR is that the join node has an updated `filterExpression` (we build an expression of inferred NOT NULL variables and `and` with the existing filter). There is not *other state change on any property of the Join node*
3.  When the iterative optimzer re-examines this JoinNode there is no 'property' or signal available that says - "this optimization was already applied"

- I could add a new property on the JoinNode, say `Optional<Boolean> hasNullsBeenInferred`, but this to me is overkill : we can acheive the same optimzation as a PlanOptimizer instead.
- I could instead try to check if the `updatedJoinFilter` is exactly the same as the existing filter and return `Result.empty` : this to me means running the rule on every JoinNode **twice**, once to apply the optimization and once to prevent re-running it on an already 'optimized' node. 

If you have ideas of how to work around this, I would be happy to update the PR

Reply:
Gentle ping @feilong-liu @kaikalur

Reply:
> I could instead try to check if the updatedJoinFilter is exactly the same as the existing filter and return Result.empty

Sounds not bad to me since the cost is small. Iterative optimizer has an advantage that it can be [logged](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java#L126) without additional work. And the constrained optimization framework also only supports [iterative optimizer](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java#L107) I think.

Reply:
> IMO this gives us fair amount of confidence that this property can be relied on.

Still good to have control on this part, maybe can do something similar [here](https://github.com/prestodb/presto/blob/9ba89b216e9f6db500389777d95ed7633c4f3d45/presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java#L324)

Reply:
@feilong-liu Change in latest iteration are

### Switched to an iterative optimizer
I worked around the issue of the iterative optimizer re-examining an already processed JoinNode by introducing a conjunct with a true expression on the final updated join filter. The rule checks for the presence of this 'marker' expression to prevent getting stuck in a loop

### Added a NotNullInferenceStrategy
Which is used to determine if a function can operate on NULL arguments. The default is `MAP_TO_STANDARD_OPERATOR`, which checks to see if the function can be mapped to a standard operator; and then uses the `OperatorType#isCalledOnNullInput` to make it's inference.  
The other strategy is `USE_FUNCTION_METADATA`, which uses `FunctionMetadata#isCalledOnNullInput`

New test to illustrate resulting behavior have been added